Unique mc evtnumber

andLaing opened the pull request:
Adds functions to generate unique event numbers for MC to allow for safe processing of split nexus events and simplify MC event mixing.

In need of more tests and a file number reader but ready for discussion.

Discussion continues from issue #715

andLaing committed:
Modify MCEventMap class to new format with evt_number and nexus_evt

andLaing committed:
Modify buffer_writer to use new style event map

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -296,6 +322,23 @@ def try_unique_evt_itr(group, itr):
     return np.unique(np.concatenate(evt_list)).astype(int)
 
 
+def load_eventnumbermap(file_name: str) -> pd.DataFrame:
```
Since you have the reader, maybe you should add a writer too instead of calling `df_writter` directly?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -108,6 +108,32 @@ def check_last_merge_index(h5out : tb.file.File) -> int:
     return -1
 
 
+def copy_nexus_eventmap(h5out   : tb.file.File,
```
Mabe add 'safe_copy ...` since in case the table is not present you don't copy and the name is a bit misleading

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -108,6 +108,32 @@ def check_last_merge_index(h5out : tb.file.File) -> int:
+    for fn in files_in:
+        evt_map = load_eventnumbermap(fn)
+        try:
+            evt_mask = evt_map.evt_number.isin(evt_arr)
```
This is a bit hard to understand. Can you add a comment that `evt_map` is None if the table is not present, and you are also protected against table present but with wrong columns?

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -108,6 +108,32 @@ def check_last_merge_index(h5out : tb.file.File) -> int:
+        except AttributeError:
+            ## No copy and return the parts of the
+            ## original arrays found.
+            if check_mc_present(fn):
```
Why do you need this check? This function should work with MC files, so the table has to be present. The `copy_mc_info` should take care it raises appropriate error/warning in case the table is not present. In case there is no MC table you probably do not even want to do any event mapping

mmkekic commented on invisible_cities/cities/components.py:
```diff
@@ -173,11 +173,12 @@ def append(l,e):
+                 event_numbers: List[int]  ,
+                 db_file      :      str   ,
+                 run_number   :      int   ,
+                 copy_eventmap: bool = True) -> None:
```
Add new parameter to docstrings

mmkekic commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -108,6 +108,32 @@ def check_last_merge_index(h5out : tb.file.File) -> int:
+                        evt_arr : np.ndarray  ,
+                        files_in: List[str]   ) -> np.ndarray:
+    """
+    Copies the nexus event map and returns the array
```
Add inputs and return to docstrings

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -104,6 +106,24 @@ def test_is_oldformat_file(ICDATADIR):
+def test_copy_nexus_eventmap(config_tmpdir, ICDATADIR):
+    file_in  = os.path.join(ICDATADIR                                  ,
+                            'nexus_new_kr83m_full.newformat.buffers.h5')
+    file_out = os.path.join(ICDATADIR, 'evtMapCopy.h5')
```
This should be a temporary path, not `ICDATADIR`

mmkekic commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -104,6 +106,24 @@ def test_is_oldformat_file(ICDATADIR):
+        assert_tables_equality(h5out .root.Run.eventMap,
+                               h5orig.root.Run.eventMap)
+
+
```
Can you add a test tho show that the map is not copied and return event are correct in case the input file does not contain the table?

andLaing committed:
Update file for buffy exact result test

andLaing committed:
Add function to copy eventmap and return nexus ids

andLaing committed:
Update components.copy_mc_info for new event ids

andLaing committed:
Update buffy for new generated IC event numbers

andLaing committed:
Add buffy with split buffers output file

To be used to test MC copy and eventMap copy
where there are split nexus events.

andLaing committed:
Add tests for split nexus event eventMap

mmkekic reviewed (approved):
This PR adds a event splitting functionality keeping unique nexus event numbers across files by assuming a constant maximum number of splits per event. A new table that maps IC event number to original MC event number is added to Run group ensuring the code is compatible with the old formats of MC production. The code is documented and tested, good job!
